### ISSUE TITLE
[Bug]: to_payload_element silently clones multimodal payloads into every request on shape mismatch — cross-request audio corruption under continuous batching

### ISSUE THREAD

**Issue by Liangtaiwan** [vllm-project/vllm-omni#4355]
### Your current environment

> Note: `collect_env.py` was run on the cluster's CPU login node, so the CUDA/GPU section reads "No CUDA". The actual serving runs on compute nodes with **NVIDIA H100 80GB** GPUs (1 GPU for the talker stage), same venv.

<details>
<summary>The output of <code>python collect_env.py</code></summary>

```text
==============================
        System Info
==============================
OS                           : Ubuntu 22.04.5 LTS (x86_64)
GCC version                  : (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version                : Could not collect
CMake version                : version 3.22.1
Libc version                 : glibc-2.35

==============================
       PyTorch Info
==============================
PyTorch version              : 2.11.0+cu126
Is debug build               : False
CUDA used to build PyTorch   : 12.6
ROCM used to build PyTorch   : N/A

==============================
      Python Environment
==============================
Python version               : 3.12.12 (main, Nov 19 2025, 22:46:53) [Clang 21.1.4 ] (64-bit runtime)
Python platform              : Linux-5.15.0-1074-azure-x86_64-with-glibc2.35

==============================
       CUDA / GPU Info
==============================
Is CUDA available            : False
CUDA runtime version         : No CUDA
CUDA_MODULE_LOADING set to   : N/A
GPU models and configuration : No CUDA
Nvidia driver version        : No CUDA
cuDNN version                : No CUDA
HIP runtime version          : N/A
MIOpen runtime version       : N/A
Is XNNPACK available         : True

==============================
          CPU Info
==============================
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        48 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               32
On-line CPU(s) list:                  0-31
Vendor ID:                            AuthenticAMD
Model name:                           AMD EPYC 7763 64-Core Processor
CPU family:                           25
Model:                                1
Thread(s) per core:                   2
Core(s) per socket:                   16
Socket(s):                            1
Stepping:                             1
BogoMIPS:                             4890.85
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl tsc_reliable nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy svm cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw topoext invpcid_single vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves clzero xsaveerptr rdpru arat npt nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload umip vaes vpclmulqdq rdpid fsrm
Virtualization:                       AMD-V
Hypervisor vendor:                    Microsoft
Virtualization type:                  full
L1d cache:                            512 KiB (16 instances)
L1i cache:                            512 KiB (16 instances)
L2 cache:                             8 MiB (16 instances)
L3 cache:                             64 MiB (2 instances)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-31
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Mitigation; safe RET, no microcode
Vulnerability Spec store bypass:      Vulnerable
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Retpolines; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

==============================
Versions of relevant libraries
==============================
[pip3] numpy==2.4.4
[pip3] nvidia-cublas==13.1.0.3
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-nvrtc-cu12==12.6.85
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cudnn-cu13==9.19.0.56
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-cufile==1.15.1.6
[pip3] nvidia-cufile-cu12==1.11.1.6
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-cusparselt-cu13==0.8.0
[pip3] nvidia-nccl-cu12==2.28.9
[pip3] nvidia-nccl-cu13==2.28.9
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvshmem-cu12==3.4.5
[pip3] nvidia-nvshmem-cu13==3.4.5
[pip3] nvidia-nvtx==13.0.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] onnxruntime==1.24.4
[pip3] torch==2.11.0
[pip3] torchaudio==2.11.0+cu126
[pip3] transformers==5.10.2
[pip3] triton==3.6.0
[conda] Could not collect

==============================
         vLLM Info
==============================
ROCM Version                 : Could not collect
vLLM Version                 : 0.20.2
vLLM-Omni Version            : 0.20.0 (git sha: ac6511f)
vLLM Build Flags:
  CUDA Archs: Not Set; ROCm: Disabled
GPU Topology:
  Could not collect

==============================
     Environment Variables
==============================
PYTORCH_NVML_BASED_CUDA_CHECK=1
TORCHINDUCTOR_COMPILE_THREADS=1
TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_cliu
LD_LIBRARY_PATH=.venv-omni2/lib/python3.12/site-packages/cv2/../../lib64:

```

</details>

### Your code version

<details>
<summary>The commit id or version of vllm</summary>

```text
vllm 0.20.2+cu129 (PyPI release)
```
</details>
<details>
<summary>The commit id or version of vllm-omni</summary>

```text
vllm-omni 0.20.0 (PyPI release)
```
</details>

### 🐛 Describe the bug

**Summary.** The per-request splitting of `multimodal_outputs` payloads slices by token offsets **only when the payload's dim-0 equals the hidden states' dim-0**; any other shape silently falls back to **cloning the whole tensor into every request's output**. For models that emit per-decode-token payloads (audio codes), this corrupts outputs under continuous batching — every request's audio accumulates other requests' frames — with no error or warning. We hit this independently of the MOSS-TTS reports (#4316 Issue 1, #4100) while serving a Moshi/Hibiki-style RQ-Transformer family (temporal + depth transformers + Mimi codec) through a `vllm.general_plugins` plugin. Payload channel: `make_omni_output → multimodal_outputs["codes"]["audio"]`, one codes row per decode token (the prompt region generates no codes). Talker stage runs with CUDA graphs enabled.

#### What happens

A payload shaped one-row-per-*decoding*-request (natural for models whose prompt generates no codes) mismatches `hidden.shape[0]` whenever:

1. a step mixes one request's **prefill chunk** with other requests' decodes (hidden rows = all scheduled tokens; payload rows = decode tokens only), or
2. **CUDA-graph padding** pads the hidden states to a capture bucket.

On mismatch, the fallback clones the full payload to every request. Under heterogeneous-length continuous batching this means cross-request audio mixing mid-utterance. We found it only by ear-testing a 4-speaker mixed batch; all structural signals (shapes, counts, exit codes) looked healthy.

#### Reproduction

Model-internal, so no copy-paste snippet — but the trigger is model-agnostic and minimal:

1. Take any talker model whose `make_omni_output` emits a payload with fewer rows than `hidden_states` (e.g. one row per decoding request, none for the prompt region).
2. Submit a batch of requests with **different prompt/output lengths**, so prefill chunks co-schedule with other requests' decode steps (identical-length lockstep batches never trigger it).
3. Inspect per-request outputs: each request's accumulated payload contains every other request's rows (whole-tensor clone), or — with truncation-style workarounds — mis-sliced neighbors / index-out-of-bounds at batch ≥ 4.

**Regression recipe that catches the whole class:** batch requests of different lengths; assert per request that (1) real payload rows == that request's generated-token count and (2) no row appears in more than one request's output. (See also the coverage note in #4100's review — identical-length batches give false confidence.)

#### Why shape-coercion workarounds are not safe

- **Truncation-alignment** (concatenate per-request payloads, truncate hidden to match — the pattern in `qwen3_tts_talker.make_omni_output`) avoids the clone but keeps the same implicit assumption: payload rows == token spans per request, in order. In mixed prefill+decode steps it gave us an index-out-of-bounds crash at batch ≥ 4; the silent variant is mis-sliced neighbors.
- **Uniform row division** (MOSS-TTS, being replaced in #4100 by `query_start_loc` recovery) misattributes rows in exactly the dangerous compositions, as that PR's review discussion documents.

#### Interim fix we ship (validated; reference implementation available)

Record each request's scheduled span at `preprocess` time (one span per request per step, batch order — no `forward_context` access at output time), and emit a **full per-token payload**: real codes rows scattered to absolute token offsets, sentinel `-1` rows over prefill spans and CUDA-graph padding (payload dim-0 == `hidden.shape[0]`, padding included). The existing splitter then routes correctly in every composition by construction; consumers drop sentinels via `codes[(codes >= 0).all(dim=1)]`. Costs: payload inflation over prefill spans (~0.5 MB/step at a 4096-token budget, 16 codebooks) and a filtering contract on consumers — but no fallback path exists to fail silently.

#### Proposed fixes

1. **Minimum (correctness)**: make the ambiguous-shape fallback in `to_payload_element` **fail loudly** (or warn once) instead of cloning. Three model families have now independently debugged silent audio corruption at this seam (MOSS-TTS in #4316/#4100; qwen3-tts implicitly via truncation; ours), and #4317 shows the same implicit-alignment disease on the cache side.
2. **Design (obsoletes all workarounds)** — either, ideally in #1601's output-channel redesign:
   - **request-keyed payload dicts** (`codes["audio"] = {req_id: tensor}`) as sketched in #4316 — no offset arithmetic exists to get wrong; or
   - **runner-provided spans**: the runner already owns `query_start_loc` (it builds `logits_indices` from it); pass the per-request spans into `make_omni_output` as an argument, so models neither fish them from `forward_context` (#4100) nor reconstruct them.

Happy to contribute our sentinel implementation and the heterogeneous-batch structural test as reference material for either direction.

Related: #4316 (Issue 1 — same routing gap, MOSS-TTS), #4100 (span-recovery workaround), #4317 (alignment corruption in the mm prefix cache), #1601 (output-channel RFC).

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

**Comment by henryj** [comment]
We hit another instance of this defect family, in qwen3-tts. `Qwen3TTSTalker.make_omni_output` collapses every request's `codes["ref"]` into one last-writer-wins variable and emits it as a length-1 list; the `element[0]` fallback in `to_payload_element` then hands that single tensor to every request in the batch. Audible result: at any concurrency > 1, every voice-clone utterance's onset decodes with another request's reference voice as context (deformed timbre converging over ~1-2 s, perfect at concurrency 1). Silent, no shape mismatch involved, so it would survive the truncation-alignment workarounds discussed above.

Filed with full diagnosis and a verified fix (batch-aligned per-request list) as #4370.

It also supports the immediate mitigation proposed here: if the ambiguous list fallback raised instead of silently reusing `element[0]`, our bug would have surfaced as an error on the first batched step instead of as corrupted audio.
